internal/wasi/fs.go

```diff
@@ -111,17 +148,159 @@ func (f *memFile) Seek(offset int64, whence int) (int64, error) {
 	case io.SeekCurrent:
 		f.offset += offset
 	case io.SeekEnd:
-		f.offset = int64(len(f.buf)) + offset
+		f.offset = int64(len(f.fsEntry.Contents)) + offset
 	default:
 		return 0, fmt.Errorf("invalid whence: %d", whence)
 	}
 	return f.offset, nil
 }
 
-// Close implements io.Closer
-func (f *memFile) Close() error {
-	if f.flush != nil {
-		f.flush(f.buf)
+// memDir represents an opened memFS directory.
+// memDir implements fs.File, fs.ReadDirFile, and io.Seeker.
+type memDir struct {
+	// name of the opened file, not a full path but a base name.
+	name string
+	// Seek offset of this opened directory.
+	offset int64
+	// Sorted file list of the dirEntries for ReadDir and Seek.
+	entries []fs.DirEntry
+	// The memFSEntry this memFileInfo is about.
+	fsEntry *memFSEntry
+}
+
+func newMemDir(dir *memFSEntry, baseName string) *memDir {
+	// Cache the file list of the directory for ReadDir to return the result in the consistent order.
+	// Note that it's ok to return stale file list if the directory is modified after Open.
+	// The result of ReadDir is undefined in POSIX in that situation, so WASI will be the same.
+	entries := make([]fs.DirEntry, 0)
+	for name, entry := range dir.Entries {
+		entries = append(entries, &memFileInfo{name: path.Base(name), fsEntry: entry})
+	}
+	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
+
+	return &memDir{
+		name:    baseName,
+		entries: entries,
+		fsEntry: dir,
 	}
+}
+
+// Stat implements fs.File.Stat
+func (d *memDir) Stat() (fs.FileInfo, error) {
+	return &memFileInfo{name: d.name, fsEntry: d.fsEntry}, nil
+}
+
+// Read implements fs.File.Read
+func (d *memDir) Read(p []byte) (int, error) {
+	return 0, fmt.Errorf("read is not supported by directory")
+}
+
+// Close implements fs.File.Close
+func (d *memDir) Close() error {
+	return nil
+}
+
+// Seek implements io.Seeker
+func (d *memDir) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
```

Comment:
should there be bounds checking to ensure d.offset doesn't end up negative?

Reply:
thanks for nice catch! Yes, there must be bounds checking. Let me add..